packages/block-editor/src/components/block-edit/edit.js

```diff
@@ -48,27 +57,223 @@ const Edit = ( props ) => {
 const EditWithFilters = withFilters( 'editor.BlockEdit' )( Edit );
 
 const EditWithGeneratedProps = ( props ) => {
-	const { attributes = {}, name } = props;
+	const { name, clientId, attributes, setAttributes } = props;
+	const registry = useRegistry();
 	const blockType = getBlockType( name );
 	const blockContext = useContext( BlockContext );
+	const registeredSources = useSelect(
+		( select ) =>
+			unlock( select( blocksStore ) ).getAllBlockBindingsSources(),
+		[]
+	);
 
-	// Assign context values using the block type's declared context needs.
-	const context = useMemo( () => {
-		return blockType && blockType.usesContext
+	const { blockBindings, context, hasPatternOverrides } = useMemo( () => {
+		// Assign context values using the block type's declared context needs.
+		const computedContext = blockType?.usesContext
 			? Object.fromEntries(
 					Object.entries( blockContext ).filter( ( [ key ] ) =>
 						blockType.usesContext.includes( key )
 					)
 			  )
 			: DEFAULT_BLOCK_CONTEXT;
-	}, [ blockType, blockContext ] );
+		// Add context requested by Block Bindings sources.
+		if ( attributes?.metadata?.bindings ) {
+			Object.values( attributes?.metadata?.bindings || {} ).forEach(
+				( binding ) => {
+					registeredSources[ binding?.source ]?.usesContext?.forEach(
+						( key ) => {
+							computedContext[ key ] = blockContext[ key ];
+						}
+					);
+				}
+			);
+		}
+		return {
+			blockBindings: replacePatternOverridesDefaultBinding(
+				name,
+				attributes?.metadata?.bindings
+			),
+			context: computedContext,
+			hasPatternOverrides: hasPatternOverridesDefaultBinding(
+				attributes?.metadata?.bindings
+			),
+		};
+	}, [
+		name,
+		blockType?.usesContext,
+		blockContext,
+		attributes?.metadata?.bindings,
+		registeredSources,
+	] );
+
+	const computedAttributes = useSelect(
+		( select ) => {
+			if ( ! blockBindings ) {
+				return attributes;
+			}
+
+			const attributesFromSources = {};
+			const blockBindingsBySource = new Map();
+
+			for ( const [ attributeName, binding ] of Object.entries(
+				blockBindings
+			) ) {
+				const { source: sourceName, args: sourceArgs } = binding;
+				const source = registeredSources[ sourceName ];
+				if ( ! source || ! canBindAttribute( name, attributeName ) ) {
+					continue;
+				}
+
+				blockBindingsBySource.set( source, {
+					...blockBindingsBySource.get( source ),
+					[ attributeName ]: {
+						args: sourceArgs,
+					},
+				} );
+			}
+
+			if ( blockBindingsBySource.size ) {
+				for ( const [ source, bindings ] of blockBindingsBySource ) {
+					// Get values in batch if the source supports it.
+					let values = {};
+					if ( ! source.getValues ) {
+						Object.keys( bindings ).forEach( ( attr ) => {
+							// Default to the the source label when `getValues` doesn't exist.
+							values[ attr ] = source.label;
+						} );
+					} else {
+						values = source.getValues( {
```

Comment:
@Mamaduka, as I'm looking at:
- https://github.com/WordPress/gutenberg/pull/67735

I'm wondering what the pattern should be for observing store changes when you expect that for the same set of dependencies, there are expected different results. In this particular case, `source.getValues` might read the value for another store, code example:

https://github.com/WordPress/gutenberg/blob/5e3d5752cca4e9005f09b1c2a8ec1a1ff89275bc/packages/editor/src/bindings/post-meta.js#L77-L89

In effect, where `coreDataStore` updates for Post Meta source, then it should compute a different result. I now realize it's a limitation of the current implementation. How can we improve it?